src/common/ina.rs

```diff
@@ -0,0 +1,246 @@
+    if registry
+        .register_power_sensor("ina219", &ina219_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+    if registry
+        .register_power_sensor("ina226", &ina226_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+}
+
+pub(crate) fn ina219_from_config(
+    cfg: ConfigType,
+    deps: Vec<Dependency>,
+) -> anyhow::Result<PowerSensorType> {
+    Ok(Arc::new(Mutex::new(from_config(Model::Ina219, cfg, deps)?)))
+}
+
+pub(crate) fn ina226_from_config(
+    cfg: ConfigType,
+    deps: Vec<Dependency>,
+) -> anyhow::Result<PowerSensorType> {
+    Ok(Arc::new(Mutex::new(from_config(Model::Ina226, cfg, deps)?)))
+}
+
+fn from_config(
+    model: Model,
+    cfg: ConfigType,
+    dependencies: Vec<Dependency>,
+) -> anyhow::Result<Ina<I2cHandleType>> {
+    let i2c_address = cfg
+        .get_attribute::<u8>("i2c_address")
+        .unwrap_or(DEFAULT_I2C_ADDRESS);
+
+    let default_max_current_amperes = match &model {
+        Model::Ina219 => 3.2,
+        Model::Ina226 => 20.0,
+    };
+    let max_current_amperes = cfg
+        .get_attribute::<f64>("max_current_amps")
+        .unwrap_or(default_max_current_amperes);
+    let max_current_nano_amperes = (max_current_amperes * 1e9) as i64;
+
+    let shunt_resistance_ohms = cfg
+        .get_attribute::<f64>("shunt_resistance")
+        .unwrap_or(DEFAULT_SHUNT_RESISTANCE_OHMS);
+    let shunt_resistance_nano_ohms = (shunt_resistance_ohms * 1e9) as i64;
+
+    let i2c_name = cfg
+        .get_attribute::<String>("i2c_bus")
+        .map_err(|_| anyhow::anyhow!("i2c_bus is a required config attribute for {}", model))?;
+    let board = get_board_from_dependencies(dependencies).ok_or(anyhow::anyhow!(
+        "a board must be configured to configure an {} sensor",
+        model
+    ))?;
+    let i2c_handle = board.get_i2c_by_name(i2c_name)?;
+
+    Ina::new(
+        model,
+        i2c_handle,
+        i2c_address,
+        max_current_nano_amperes,
+        shunt_resistance_nano_ohms,
+    )
+}
+
+#[derive(Clone, Copy)]
+pub enum Model {
+    Ina219,
+    Ina226,
+}
+
+impl fmt::Display for Model {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let model_str = match self {
+            Model::Ina219 => "INA219",
+            Model::Ina226 => "INA226",
+        };
+        write!(f, "{model_str}")
+    }
+}
+
+pub struct Ina<H> {
+    model: Model,
+    i2c_handle: H,
+    i2c_address: u8,
+    calibration_scale_bytes: [u8; 2],
+    max_current_nano_amperes: i64,
+    power_reading_lsb: i64,
+}
+
+impl<H> Ina<H>
+where
+    H: I2CHandle,
+{
+    fn new(
+        model: Model,
+        i2c_handle: H,
+        i2c_address: u8,
+        max_current_nano_amperes: i64,
+        shunt_resistance_nano_ohms: i64,
+    ) -> anyhow::Result<Self> {
+        let current_reading_lsb = max_current_nano_amperes / (1 << 15);
```

Comment:
How carefully do you want me to review the math and logic that connects to the sensor specs? Ditto register usage, etc. Or was this largely cribbed from another ina sensor impl so you have good confidence in your translation to micro-rdk?

Reply:
OK, so I will not go over the detailed logic here too carefully.